### PR TITLE
Engine State Addresses

### PR DESCRIPTION
**Pull request by passabilities**
The EngineState is given a significant upgrade in terms of processing and speed. 

### Merge tx query and subscribe code paths
Functions have been stripped out to be modular so they can be called at any time. This allows the same functions to be called when loading addresses from the database to process and when an address that is being watched triggers a new transaction.

### getFreshAddress & addGapLimitAddresses
Functions have been updated in `edge-core-js` to allow promises to be returned. Adding gap limit addresses functionality added by allowing addresses to be saved to the database without being linked to a path. In order to make this possible, the structure of the address databases needed to change to store all the data for the address by scriptPubkey rather than by the path.

### CODE REVIEW

File: src/common/utxobased/engine/makeUtxoEngineState.ts [hunk 1]
```diff
@@ -47,299 +59,567 @@ export function makeUtxoEngineState(config: UtxoEngineStateConfig): UtxoEngineSt
...
+  }
 
-    if (accountDetails.page < accountDetails.totalPages) {
-      await processAddressTransactions(address, ++page)
-    }
+  const { scriptPubkey } = walletTools.getScriptPubkey(fromPath)
+  const addressData = await processor.fetchAddressByScriptPubkey(scriptPubkey)
+  if (addressData?.used) {
+    fromPath.addressIndex++
+    return findFreshIndex(args)
```
Comment (passabilities) [inline src/common/utxobased/engine/makeUtxoEngineState.ts:330]: if the previous index is not used traverse backwards

File: src/common/utxobased/engine/makeUtxoEngineState.ts [hunk 2]
```diff
@@ -47,299 +59,567 @@ export function makeUtxoEngineState(config: UtxoEngineStateConfig): UtxoEngineSt
...
+      const changeIndex = change ? 1 : 0
+      if (walletPurpose === BIP43PurposeTypeEnum.Segwit) {
+        const segwitFormat = getCurrencyFormatFromPurposeType(BIP43PurposeTypeEnum.Segwit)
+        const segwitIndex = await getFreshIndex({
+          ...config,
+          format: segwitFormat,
+          changeIndex,
+          find: false
+        })
+        const { address: segwitAddress } = walletTools.getAddress({
```
Comment (passabilities) [inline src/common/utxobased/engine/makeUtxoEngineState.ts:106]: Get the address from the script pubkey from the processor

File: src/common/utxobased/engine/makeUtxoEngineState.ts [hunk 3]
```diff
@@ -47,299 +59,567 @@ export function makeUtxoEngineState(config: UtxoEngineStateConfig): UtxoEngineSt
...
       })
+      await Promise.all(addNewPromises)
+    },
 
-      gap = address.used ? 0 : gap + 1
-      path = {
-        ...path,
-        addressIndex: path.addressIndex + 1
-      }
+    async markAddressUsed(address: string): Promise<void> {
```
Comment (passabilities) [inline src/common/utxobased/engine/makeUtxoEngineState.ts:161]: implement

**Comment by thehobbit85** [comment]
I'm guessing this is irrelevant since we are restructuring the PR

**Pull request closed without merging**